akka-http-core/src/main/scala/akka/http/scaladsl/ConnectionContext.scala

```diff
@@ -23,15 +28,60 @@ trait ConnectionContext extends akka.http.javadsl.ConnectionContext {
 object ConnectionContext {
   // ConnectionContext
   //#https-context-creation
+  def https(createSSLEngine: Option[(String, Int)] => () => SSLEngine) =
+    new HttpsConnectionContext(Right(o => Engine(createSSLEngine(o))))
+
+  def httpsServer(sslContext: SSLContext) =
+    new HttpsConnectionContext(Right({
+      case None => Engine(() => {
+        val engine = sslContext.createSSLEngine()
+        engine.setUseClientMode(false)
+        engine
+      })
+      case Some((host, port)) => Engine(() => {
+        val engine = sslContext.createSSLEngine(host, port)
+        engine.setUseClientMode(false)
+        engine
+      })
+    }))
+
+  def httpsClient(context: SSLContext)(implicit system: ActorSystem) = {
+    val verifier = new DefaultHostnameVerifier(new AkkaLoggerFactory(system))
+
+    new HttpsConnectionContext(Right {
+      case None =>
+        Engine(() => {
+          val engine = context.createSSLEngine()
+          engine.getSSLParameters.setEndpointIdentificationAlgorithm("https")
+          engine.setUseClientMode(true)
+          engine
+        })
+      case Some((host, port)) =>
+        Engine(
+          () => {
+            val engine = context.createSSLEngine(host, port)
+            engine.getSSLParameters.setEndpointIdentificationAlgorithm("https")
+            engine.getSSLParameters.setServerNames(Collections.singletonList(new SNIHostName(host)))
```

Comment:
As mentioned in #3255, making SNI enabled by default here and above is an important feature, just commenting here for posterity so there's a record of this decision being very intentional in the PR.

Reply:
Do you mean we should enable it explicitly? Usually, it isn't necessary to set it explicitly when using `context.createSSLEngine(host, port)`.

Reply:
Hmm, it does seem so (in TlsEndpointVerificationSpec)

Reply:
I just tried this both with jdk8 and jdk11, and indeed the SNI seems to be set even when this parameter is not explicitly set.

It seems to have been very deliberately set in https://github.com/akka/akka/pull/19858/files#diff-d2926806d32db424bb9b3e288fc1fb0cR159 though - which was also already using `createSSLEngine(hostname, port)`.

Reply:
Verified it is set correctly automatically